### PR TITLE
Handling Loss of Synchronization due to Significant Packet Loss (RFC4303 Appendix A3)

### PR DESCRIPTION
**Pull request by vanfstd**
This implements and adds tests for RFC4303 Appendix A3, re-synchronization of the anti-replay mechanism in the event of massive packet loss.

### CODE REVIEW

File: src/lib/ipsec/esp.lua [hunk 1]
```diff
@@ -168,10 +162,59 @@ function esp_v6_decrypt:decapsulate (p)
                    "flow_id=" .. tostring(self.ip:flow_label()) .. ", " ..
                    "reason='" .. reason .. "'";
       logger:log("Rejecting packet ("..info..")")
-      return false
+      -- RFC4303 is somewhat unclear on a) whether or not packets that look replayed are eligible to trigger
+      -- recyns, and b) what 'consecutive' means in "number of consecutive packets that fail authentication":
+      -- Consecutive in time? Consecutive in their seq_low's?
+      -- Assuming a) should be 'yes' since the to-be-resynced packets might have seq_low's that just happen
+      -- to fall inside the replay window (seq_lo-wise) and would look replayed, yet aren't.
+      -- Assuming b) means 'in time', since systematic loss could stall resync indefinitely.
```
Comment (eugeneia) [inline src/lib/ipsec/esp.lua]: Do we know how other implementations interpret the RFC? If not, could you compare with a mature implementation? OpenBSD?

File: src/lib/ipsec/esp.lua [hunk 2]
```diff
@@ -168,10 +162,59 @@ function esp_v6_decrypt:decapsulate (p)
...
+         end
+         self.decap_fail = 0 -- avoid immediate re-triggering if resync fails
+         seq_high_tmp = self:resync(p, seq_low, seq_high_tmp + 1, ESP_RESYNC_ATTEMPTS)
+         if seq_high_tmp >= 0 then
+            seq_high = seq_high_tmp
+            proceed = true -- resynced! the data has been decrypted in the process so we're ready to go
+         end
+      end
    end
+   if not proceed then return false end
```
Comment (eugeneia) [inline src/lib/ipsec/esp.lua]: Looks to me as if we could `return false` in the failure branch if resync failed and avoid the `proceed` variable altogether?

File: src/lib/ipsec/esp.lua [hunk 3]
```diff
@@ -48,6 +48,8 @@ local PAYLOAD_OFFSET = ETHERNET_SIZE + IPV6_SIZE
 local ESP_NH = 50 -- https://tools.ietf.org/html/rfc4303#section-2
 local ESP_SIZE = esp:sizeof()
 local ESP_TAIL_SIZE = esp_tail:sizeof()
+local ESP_RESYNC_THRESHOLD = 4 -- Resync after this many consecutive failed decaps
+local ESP_RESYNC_ATTEMPTS = 6 -- Try this many increments of seq_high
```
Comment (eugeneia) [inline src/lib/ipsec/esp.lua]: These should be configurable in `new` like spi etc. What could be sensible defaults?

File: src/lib/ipsec/esp.lua (hunk 2 above)
Comment (vanfstd) [inline src/lib/ipsec/esp.lua]: We could drop `proceed` at the cost of two `else return false`, yes.  Good catch.

File: src/lib/ipsec/esp.lua [hunk 4]
```diff
@@ -168,10 +161,61 @@ function esp_v6_decrypt:decapsulate (p)
...
-      return false
+      -- RFC4303 is somewhat unclear on a) whether or not packets that look replayed are eligible to trigger
+      -- recyns, and b) what 'consecutive' means in "number of consecutive packets that fail authentication":
+      -- Consecutive in time? Consecutive in their seq_low's?
+      -- Assuming a) should be 'yes' since the to-be-resynced packets might have seq_low's that just happen
+      -- to fall inside the replay window (seq_lo-wise) and would look replayed, yet aren't.
+      -- Assuming b) means 'in time', since systematic loss could stall resync indefinitely.
+      self.decap_fail = self.decap_fail + 1
+      if self.decap_fail >= ESP_RESYNC_THRESHOLD then
+         local seq_high_tmp = seq_high
```
Comment (vanfstd) [inline src/lib/ipsec/esp.lua]: Rename to `resync_start`?

File: src/lib/ipsec/esp.lua [hunk 5]
```diff
@@ -168,10 +161,61 @@ function esp_v6_decrypt:decapsulate (p)
...
+   local gcm = self.aes_128_gcm
+   local payload = p.data + PAYLOAD_OFFSET
+   local iv_start = payload + ESP_SIZE
+   local ctext_start = payload + self.CTEXT_OFFSET
+   local ctext_length = p.length - self.PLAIN_OVERHEAD
+   for i = 1, n do
+      if gcm:decrypt(ctext_start, seq_low, seq_high, iv_start, ctext_start, ctext_length) then
+         return seq_high
+      else
+         C.memmove(p.data, p_orig.data, p_orig.length)
```
Comment (vanfstd) [inline src/lib/ipsec/esp.lua]: clone/free instead?

File: src/lib/ipsec/esp.lua (hunk 3 above)
Comment (vanfstd) [inline src/lib/ipsec/esp.lua]: Some considerations: a) Anyone who can inject or forge packets might be able to trigger an attempt to resynchronize by sending packets with the right (unencrypted) header fields or replaying very old packets.  Every `SP_RESYNC_THRESHOLD`th such packet will trigger resync (as opposed to *every* such packet after the `SP_RESYNC_THRESHOLD`th due to [Line 180](https://github.com/vanfstd/snabb/blob/ef2b4027d3baaabf337e8fb521d3bdd76f994b15/src/lib/ipsec/esp.lua#L180))

Next, the actual cost of a resync operation is roughly `ESP_RESYNC_ATTEMPTS` times the cost of decrypting a packet payload. That's not much in absolute terms, but compared to the overhead of regular packets, it is.  However, the larger the value of `ESP_RESYNC_ATTEMPTS`, the larger a packet loss we can deal with before losing the ability to resynchronize. The number of lost packets we can tolerate is 2^32 times `ESP_RESYNC_ATTEMPTS`.

Based on that, I'd tend to default to a smallish value for `ESP_RESYNC_ATTEMPTS`. Even a value as low as 5 would result in being able to resynchronize after well over a year of network outage, if the sender keeps up a packet rate of 1000 p/s. We could bump it to 6 or even 8 to feel extra sure about it.

Contrary, the value of `ESP_RESYNC_THRESHOLD` should probably default to something substantially bigger. Maybe 1024.  If we are in a situation where we'd resync, we're already past 2^32 lost packets, so it doesn't matter if we require some more to fail before spending resources on resync, making it less attractive to try and abuse resync for DoS purposes in the process.

File: src/lib/ipsec/esp.lua [hunk 6]
```diff
@@ -168,18 +161,72 @@ function esp_v6_decrypt:decapsulate (p)
...
+            gcm:encrypt(ctext_start, iv_start, seq_low, seq_high, ctext_start, ctext_length, ffi.new("uint8_t[?]", gcm.AUTH_SIZE))
+            resync_start = seq_high + 1
+         else
+            resync_start = self:seq_high() + 1 -- use the last seq_high we saw if it looked replayed
+         end
+         self.decap_fail = 0 -- avoid immediate re-triggering if resync fails
+         local seq_high_resynced = self:resync(p, seq_low, resync_start, self.resync_attempts)
+         if seq_high_resynced >= 0 then
+            seq_high = seq_high_resynced
+            -- resynced! the data has been decrypted in the process so we're ready to go
```
Comment (eugeneia) [inline src/lib/ipsec/esp.lua]: I just wondered: what happens to the state stored in the anti-replay window on resync? Doesn't it need to be updated/cleared to reflect the advance in `seq_high`?

Comment (vanfstd) [inline src/lib/ipsec/esp.lua]: That is being taken care of by the call to track_seq_no a few lines below this. Effectively it calls [advance_window](https://github.com/vanfstd/snabb/blob/ef2b4027d3baaabf337e8fb521d3bdd76f994b15/src/lib/ipsec/track_seq_no.c#L44) with the resynced-on sequence number, resulting in `diff` in that function to be large so the entire window is cleared.
Might still make sense to add a test for this, though.

**Comment by eugeneia** [comment]
We decided to merge this with some refactoring, and additional bug fixes (RFC4303 never gets old):

- [Hydra report](https://hydra.snabb.co/build/599710/download/2/report.html#density-plot) (`resync-reorg` is what was merged)
- [Diff of resync-reorg](https://github.com/vanfstd/snabb/compare/resync-pr...eugeneia:resync-reorg)

**Pull request merged**